File: babbleapp/src/main/java/com/learning/leap/bwb/userInfo/UserInfoPresenter.java
package com.learning.leap.bwb.userInfo;


import com.amazonaws.mobileconnectors.dynamodbv2.dynamodbmapper.DynamoDBMapper;
import com.learning.leap.bwb.BuildConfig;
import com.learning.leap.bwb.Player;
import com.learning.leap.bwb.R;
import com.learning.leap.bwb.helper.LocalLoadSaveHelper;
import com.learning.leap.bwb.research.ResearchNotifications;
import com.learning.leap.bwb.research.ResearchPlayers;
import com.learning.leap.bwb.utility.NetworkCheckerInterface;
import com.learning.leap.bwb.models.BabblePlayer;
import com.learning.leap.bwb.models.Notification;

import java.util.List;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import io.realm.Realm;

public class UserInfoPresenter{
   private UserInfoViewInterface userInfoViewInterface;
    private boolean newUser;
   private Player babblePlayer;
    private final CompositeDisposable disposables = new CompositeDisposable();
    private NetworkCheckerInterface networkCheckerInterface;
    private LocalLoadSaveHelper saveHelper;
    private Realm realm;
    private DynamoDBMapper mapper;

    public UserInfoPresenter(Boolean newUser, UserInfoViewInterface userInfoViewInterface, LocalLoadSaveHelper saveHelper, NetworkCheckerInterface networkCheckerInterface, Realm realm, DynamoDBMapper mapper){
        this.newUser = newUser;
        this.userInfoViewInterface = userInfoViewInterface;
        this.networkCheckerInterface = networkCheckerInterface;
        this.saveHelper = saveHelper;
        this.realm = realm;
        this.mapper = mapper;
    }


    public void onDestory() {
        disposables.clear();

    }

    private void updatePlayer() {
         disposables.add(babblePlayer.savePlayerObservable(mapper,networkCheckerInterface,saveHelper).subscribe(c -> updatePlayerOnCompleted(),
                 throwable -> userInfoViewInterface.displayErrorDialog(R.string.BabbleError, R.string.userSave)));
    }

    private void updatePlayerOnCompleted(){
        if (networkCheckerInterface.isConnected()) {
            if (newUser) {
                retriveNotificationsFromAmazon();
            } else {
                userInfoViewInterface.dismissActivity();
            }
        }else{
            userInfoViewInterface.displayErrorDialog(R.string.BabbleError,R.string.noConnection);
        }
    }


    private void retriveNotificationsFromAmazon() {
            Disposable notificationDisposable = babblePlayer.retriveNotifications(babblePlayer.getuserAgeInMonth(), mapper)
                    .doOnSubscribe(disposable -> userInfoViewInterface.displaySaveDialog())
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(this::updateViewAfterRetrivingNotificationList, throwable -> updateViewAfterError());
            disposables.add(notificationDisposable);

    }

    private void updateViewAfterError(){
        userInfoViewInterface.dismissSaveDialog();
        userInfoViewInterface.displayErrorDialog(R.string.BabbleError,R.string.downloadError);
    }

    private void updateViewAfterRetrivingNotificationList(List<Notification>notifications){
        userInfoViewInterface.dismissSaveDialog();
        if (notifications == null || notifications.size() == 0){
            userInfoViewInterface.displayErrorDialog(R.string.BabbleError,R.string.noPromptsForUser);
        }else {
            saveNotifications(notifications);
        }
    }

    private void updateNWViewAfterRetrivingNotificationList(List<ResearchNotifications>notifications){
        userInfoViewInterface.dismissSaveDialog();
        if (notifications == null || notifications.size() == 0){
            userInfoViewInterface.displayErrorDialog(R.string.BabbleError,R.string.noPromptsForUser);
        }else {
            saveNWNotifications(notifications);
        }
    }

    private void saveNotifications(List<Notification> notifications) {
        for (int i = 0; i< notifications.size(); i++){
            notifications.get(i).setId(i);
        }
        saveHelper.saveNotificationSize(notifications.size());
        realm.beginTransaction();
        realm.where(Notification.class).findAll().deleteAllFromRealm();
        realm.copyToRealm(notifications);
        babblePlayer.savePlayerToRealm();
        realm.commitTransaction();
        userInfoViewInterface.downloadIntent();
    }

    private void saveNWNotifications(List<ResearchNotifications> notifications) {
        for (int i = 0; i< notifications.size(); i++){
            notifications.get(i).setId(i);
        }
        saveHelper.saveNotificationSize(notifications.size());
        realm.beginTransaction();
        realm.where(Notification.class).findAll().deleteAllFromRealm();
        realm.copyToRealm(notifications);
        babblePlayer.savePlayerToRealm();
        realm.commitTransaction();
        userInfoViewInterface.downloadIntent();
    }


    public void createBabblePlayer(Player babblePlayer) {
        this.babblePlayer = babblePlayer;
    }

    public void loadPlayerFromSharedPref() {
        if (!BuildConfig.FLAVOR.equals("regular")) {
            ResearchPlayers researchPlayers = new ResearchPlayers();
            babblePlayer = researchPlayers.loadBabblePlayerFronSharedPref(saveHelper);
            userInfoViewInterface.displayUserInfo(babblePlayer);
        }else {
            BabblePlayer localBabblePlayer = new BabblePlayer();
            babblePlayer = localBabblePlayer.loadBabblePlayerFronSharedPref(saveHelper);
            userInfoViewInterface.displayUserInfo(babblePlayer);
        }

    }


    public void checkUserInput() {
        babblePlayer.setuserAgeInMonth();
        if (babblePlayer.checkIfPlayerIsValid()){
            updatePlayer();
        }else if (!babblePlayer.checkZipCode()){
            userInfoViewInterface.displayErrorDialog(R.string.userZipCodeErrorTitle,R.string.userZipCodeError);
        }else if (babblePlayer.checkNameIsEmpty()){
            userInfoViewInterface.displayErrorDialog(R.string.userNameNameErrorTitle,R.string.userNameEmptyError);
        }else if (babblePlayer.checkNameIsTooLong()) {
            userInfoViewInterface.displayErrorDialog(R.string.userNameNameErrorTitle,R.string.userNameToLongError);
        }else if (!babblePlayer.checkDate()){
            userInfoViewInterface.displayErrorDialog(R.string.userBirthdayErrorTitle,R.string.userBirthdayError);
        }else if (babblePlayer.getuserAgeInMonth() <= 0){
            userInfoViewInterface.displayErrorDialog(R.string.BabbleError,R.string.userNotBornYetError);
        }
    }

}

File: babbleapp/src/main/java/com/learning/leap/bwb/userInfo/UserInfoPresenter.kt
package com.learning.leap.bwb.userInfo

import com.amazonaws.mobileconnectors.dynamodbv2.dynamodbmapper.DynamoDBMapper
import com.learning.leap.bwb.R
import com.learning.leap.bwb.helper.LocalLoadSaveHelper
import com.learning.leap.bwb.model.BabbleTip
import com.learning.leap.bwb.model.BabbleUser
import com.learning.leap.bwb.models.BabblePlayer
import com.learning.leap.bwb.models.Notification
import com.learning.leap.bwb.research.ResearchNotifications
import com.learning.leap.bwb.utility.NetworkCheckerInterface
import io.reactivex.Completable
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import io.realm.Realm

class UserInfoPresenter(private val newUser: Boolean, private val userInfoViewInterface: UserInfoViewInterface, private val saveHelper: LocalLoadSaveHelper, private val networkCheckerInterface: NetworkCheckerInterface, private val realm: Realm, private val mapper: DynamoDBMapper) {
    private var babblePlayer: BabbleUser? = null
    private val disposables = CompositeDisposable()

    fun onDestroy() {
        disposables.clear()
    }

    private fun updatePlayer() {
        babblePlayer?.let {
            val disposable =   Completable.fromAction {
                if (networkCheckerInterface.isConnected) {
                    it.saveBabbleUser(mapper, saveHelper)
                } else {
                    throw java.lang.Exception()
                }
            }.subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
                    .subscribe({
                        updatePlayerOnCompleted()
                    }, {
                        it.printStackTrace()
                        userInfoViewInterface.displayErrorDialog(R.string.BabbleError, R.string.userSave)
                    })
            disposables.add(disposable)
        }
    }

    private fun updatePlayerOnCompleted() {
        if (networkCheckerInterface.isConnected) {
            if (newUser) {
                retrieveNotificationsFromAmazon()
            } else {
                userInfoViewInterface.dismissActivity()
            }
        } else {
            userInfoViewInterface.displayErrorDialog(R.string.BabbleError, R.string.noConnection)
        }
    }

    private fun retrieveNotificationsFromAmazon() {
        babblePlayer?.let {
            Single.fromCallable {
                return@fromCallable it.retrieveNotifications(it.userAgeInMonth, mapper)
            }.doOnSubscribe { userInfoViewInterface.displaySaveDialog() }
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe({ tips ->
                        updateViewAfterRetrievingNotificationList(tips)
                    }, {
                        it.printStackTrace()
                            updateViewAfterError()
                    })
        }
    }

    private fun updateViewAfterError() {
        userInfoViewInterface.dismissSaveDialog()
        userInfoViewInterface.displayErrorDialog(R.string.BabbleError, R.string.downloadError)
    }

    private fun updateViewAfterRetrievingNotificationList(notifications: List<BabbleTip>) {
        userInfoViewInterface.dismissSaveDialog()
        if (notifications.isEmpty()) {
            userInfoViewInterface.displayErrorDialog(R.string.BabbleError, R.string.noPromptsForUser)
        } else {
            saveNotifications(notifications)
        }
    }
    private fun saveNotifications(notifications: List<BabbleTip>) {
        for (i in notifications.indices) {
            notifications[i].id = i
        }
        saveHelper.saveNotificationSize(notifications.size)
        realm.beginTransaction()
        realm.where(Notification::class.java).findAll().deleteAllFromRealm()
        realm.copyToRealm(notifications)
        babblePlayer?.savePlayerToRealm(realm)
        realm.commitTransaction()
        userInfoViewInterface.downloadIntent()
    }


    fun createBabblePlayer(user: BabbleUser?) {
        babblePlayer = user
    }

    fun loadPlayerFromSharedPref() {
        babblePlayer = BabbleUser.loadBabblePlayerFromSharedPref(saveHelper)
        userInfoViewInterface.displayUserInfo(babblePlayer!!)
    }

    fun checkUserInput() {
        babblePlayer?.let {
            if (it.checkIfPlayerIsValid()!!) {
                updatePlayer()
            } else if (it.checkNameIsEmpty()!!) {
                userInfoViewInterface.displayErrorDialog(R.string.userNameNameErrorTitle, R.string.userNameEmptyError)
            } else if (it.checkNameIsTooLong()!!) {
                userInfoViewInterface.displayErrorDialog(R.string.userNameNameErrorTitle, R.string.userNameToLongError)
            } else if (!it.checkDate()) {
                userInfoViewInterface.displayErrorDialog(R.string.userBirthdayErrorTitle, R.string.userBirthdayError)
            } else if (it.userAgeInMonth <= 0) {
                userInfoViewInterface.displayErrorDialog(R.string.BabbleError, R.string.userNotBornYetError)
            }
        }

    }

}